### PR TITLE
`gcs-upload`: add a couple of enhancements to upload compression logic

### PR DESCRIPTION
**Pull request by smg247**
There are a few items that we should check prior to compressing a file:
- If files are too small, don't compress. 
- If files are already gzip, don't compress despite extension. (this case occurs due to https://github.com/kubernetes/test-infra/blob/9be1ab79c2696e56b604a3304d5e7be911f018dd/prow/pod-utils/gcs/metadata.go#L27-L71)
- Never compress 'gz' or 'gzip' files even when compressing '*'.

### CODE REVIEW

File: prow/pod-utils/gcs/upload.go [hunk 1]
```diff
@@ -65,14 +65,57 @@ func Upload(ctx context.Context, bucket, gcsCredentialsFile, s3CredentialsFile s
...
+	}
+
+	file, err := os.Open(dest)
+	if err != nil {
+		logrus.WithError(err).Warnf("unable to open file: %s to determine if we should compress", dest)
+		return false
+	}
+	defer file.Close()
+
+	header := make([]byte, 2)
```
Comment (stevekuznetsov) [inline prow/pod-utils/gcs/upload.go:101]: Let's use `http.DetectContentType()` so we don't have to write and own this logic.

**Review by stevekuznetsov — COMMENTED** [review]
Given that opening the files is only for detecting gzipped files that are not given the correct file extension, I wonder if there's some utility to making that step optional. I'd expect the frequency of gzipped-but-not-correctly-labelled files to be low. Naively I'd also expect `*` to be the most common compression filter - have you been able to measure the runtime impact of having to open every file in the artifact tree twice?

**Comment by smg247** [comment]
> Given that opening the files is only for detecting gzipped files that are not given the correct file extension, I wonder if there's some utility to making that step optional. I'd expect the frequency of gzipped-but-not-correctly-labelled files to be low.

It actually happens all the time, at least in the `OpenShift` installation. We zip log files, and save them as `some-log.log.gz`, and then the logic linked in the description removes the `.gz` extension.

> Naively I'd also expect `*` to be the most common compression filter - have you been able to measure the runtime impact of having to open every file in the artifact tree twice?

No, I haven't been able to think of a reasonable way to do so.

**Comment by smg247** [comment]
I am not seeing any metrics, but there is some decent logging that we could use the time delta on to determine if we are adding an unreasonable amount of processing time here. I will check this out post-merge.

**Comment by stevekuznetsov** [comment]
@smg247 the Kubernetes API server metrics for containers transitioning between states might be useful for this

**Comment by stevekuznetsov** [comment]
`gcsupload`? :D - but yeah, that's what I was thinking.

**Comment by stevekuznetsov** [comment]
Ah right ... `sidecar` runs the whole time. :thinking:

**Pull request merged**